src/test/java/com/atlassian/bitbucket/jenkins/internal/fixture/JenkinsProjectRule.java

```diff
@@ -0,0 +1,173 @@
+package com.atlassian.bitbucket.jenkins.internal.fixture;
+
+import hudson.model.FreeStyleProject;
+import hudson.model.Job;
+import hudson.scm.SCM;
+import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * A test rule designed to be used as part of a test chain to provide your tests with the ability to run against multiple
+ * kinds of supported Jenkins projects, and ensure your features have compatibility without having to write a separate
+ * test case for each one. The rule is designed for black-box testing (treating any kind of Jenkins project as a generic
+ * {@link hudson.model.Job}), and can be customized to run with any chosen set of projects if your feature is not
+ * intended to have full compatibility. This rule can also be used to handle project instantiation and cleanup for individual jobs.
+ */
+public class JenkinsProjectRule implements TestRule {
+
+    // The set of all project types that can be tested by this rule, specified in the builder
+    private final Set<ProjectType> testedProjectTypes;
+    // The chained JenkinsRule, used to instantiate projects
+    private final JenkinsRule jenkinsRule;
```

Comment:
This is a kind of frustrating workaround where we need access to the Jenkins object to instantiate the projects (before the test has necessarily run, as `before` is only performed during the evaluate phase, when the tests have already been created). Extending the JenkinsProjectRule gets the order wrong- as this becomes the outer rule in the chain, when we can't assume the jenkins object has been instantiated yet.
Open to ideas on how we can avoid this.

Reply:
It seems fine to me, and as discussed in person, I prefer aggregation/composition to inheritance whenever possible.